util/rateLimit.go

```diff
@@ -0,0 +1,470 @@
+	}
+	// make capacity for the provided client
+	q := capacityQueue(make(chan capacity, erl.CapacityPerReservation))
+	erl.capacityByClient[c] = q
+	erl.clientLock.Unlock()
+
+	// create a thread to drain the capacity from sharedCapacity in a blocking way
+	// and move it to the reservation, also in a blocking way
+	go func() {
+		for i := 0; i < erl.CapacityPerReservation; i++ {
+			erl.sharedCapacity.blockingConsume()
+			q.blockingRelease()
+		}
+	}()
+	return q, nil
+}
+
+// closeReservation will remove the client mapping to capacity channel,
+// and will kick off a routine to drain the capacity and replace it to the shared capacity
+func (erl *ElasticRateLimiter) closeReservation(c ErlClient) {
+	erl.clientLock.Lock()
+	q, exists := erl.capacityByClient[c]
+	// guard clauses, and preventing the ElasticRateLimiter from draining its own sharedCapacity
+	if !exists || q == erl.sharedCapacity {
+		erl.clientLock.Unlock()
+		return
+	}
+	delete(erl.capacityByClient, c)
+	erl.clientLock.Unlock()
+
+	// start a routine to consume capacity from the closed reservation, and return it to the sharedCapacity
+	go func() {
+		for i := 0; i < erl.CapacityPerReservation; i++ {
+			q.blockingConsume()
+			erl.sharedCapacity.blockingRelease()
+		}
+	}()
+}
+
+// CongestionManager is an interface for tracking events which happen to capacityQueues
+type CongestionManager interface {
+	Start(ctx context.Context, wg *sync.WaitGroup)
+	Consumed(c ErlClient, t time.Time)
+	Served(t time.Time)
+	ShouldDrop(c ErlClient) bool
+}
+
+type event struct {
+	c ErlClient
+	t time.Time
+}
+
+type shouldDropQuery struct {
+	c   ErlClient
+	ret chan bool
+}
+
+// "Random Early Detection" congestion manager,
+// will propose to drop messages proportional to the caller's request rate vs Average Service Rate
+type redCongestionManager struct {
+	runLock                *deadlock.Mutex
+	running                bool
+	window                 time.Duration
+	consumed               chan event
+	served                 chan event
+	shouldDropQueries      chan shouldDropQuery
+	targetRate             float64
+	targetRateRefreshTicks int
+	// exp is applied as an exponential factor in shouldDrop. 1 would be linearly proportional, higher values punish noisy neighbors more
+	exp float64
+	// consumed is the only value tracked by-queue. The others are calculated in-total
+	// TODO: If we desire later, we can add mappings onto release/done for more insight
+	consumedByClient map[ErlClient]*[]time.Time
+	serves           []time.Time
+}
+
+// NewREDCongestionManager creates a Congestion Manager which will watches capacityGuard activity,
+// and regularly calculates a Target Service Rate, and can give "Should Drop" suggestions
+func NewREDCongestionManager(d time.Duration, r int) *redCongestionManager {
+	ret := redCongestionManager{
+		runLock:                &deadlock.Mutex{},
+		window:                 d,
+		consumed:               make(chan event, 100000),
```

Comment:
100k?

Reply:
Talked offline -- going to set this to the same as the txBacklog size, which would allow for the whole capacity of the queue to emit events. Might need tuning later.